usersession/client/client.go

```diff
@@ -0,0 +1,244 @@
+			uidStr := filepath.Base(filepath.Dir(socket))
+			uid, err := strconv.Atoi(uidStr)
+			if err != nil {
+				logger.Noticef("Socket %q does not appear to be in a valid XDG_RUNTIME_DIR", socket)
+				return
+			}
+			u := url.URL{
+				Scheme:   "http",
+				Host:     uidStr,
+				Path:     urlpath,
+				RawQuery: query.Encode(),
+			}
+			req, err := http.NewRequest(method, u.String(), bytes.NewBuffer(body))
+			if err != nil {
+				logger.Noticef("Failed to create HTTP request: %v", err)
+				return
+			}
+			req = req.WithContext(ctx)
+			for key, value := range headers {
+				req.Header.Set(key, value)
+			}
+			httpResp, err := client.doer.Do(req)
+			response := response{
+				uid: uid,
+				err: err,
+			}
+			if err == nil {
+				defer httpResp.Body.Close()
+				response.statusCode = httpResp.StatusCode
+				response.err = decodeInto(httpResp.Body, &response)
+				response.checkError()
+			}
+			mu.Lock()
+			defer mu.Unlock()
+			responses = append(responses, &response)
+		}(socket)
+	}
+	wg.Wait()
+	return responses, nil
+}
+
+func decodeInto(reader io.Reader, v interface{}) error {
+	dec := json.NewDecoder(reader)
+	if err := dec.Decode(v); err != nil {
+		r := dec.Buffered()
+		buf, err1 := ioutil.ReadAll(r)
+		if err1 != nil {
+			buf = []byte(fmt.Sprintf("error reading buffered response body: %s", err1))
+		}
+		return fmt.Errorf("cannot decode %q: %s", buf, err)
+	}
+	return nil
+}
+
+type SessionInfo struct {
+	Version string `json:"version"`
+}
+
+func (client *Client) SessionInfo(ctx context.Context) (info map[int]SessionInfo, err error) {
+	responses, err := client.do(ctx, "GET", "/v1/session-info", nil, nil, nil)
+	if err != nil {
+		return nil, err
+	}
+
+	info = make(map[int]SessionInfo)
+	for _, resp := range responses {
+		if resp.err != nil {
+			if err == nil {
+				err = resp.err
+			}
+			continue
+		}
+		var si SessionInfo
+		if decodeErr := json.Unmarshal(resp.Result, &si); decodeErr != nil {
+			if err == nil {
+				err = decodeErr
+			}
+			continue
+		}
+		info[resp.uid] = si
+	}
+	return info, err
+}
+
+type ServiceFailure struct {
+	Uid     int
+	Service string
+	Error   string
+}
+
+func (client *Client) servicesCall(ctx context.Context, action string, services []string) (failures []ServiceFailure, err error) {
+	headers := map[string]string{"Content-Type": "application/json"}
+	reqBody, err := json.Marshal(map[string]interface{}{
+		"action":   action,
+		"services": services,
+	})
+	if err != nil {
+		return nil, err
+	}
+	responses, err := client.do(ctx, "POST", "/v1/services", nil, headers, reqBody)
+	if err != nil {
+		return nil, err
+	}
+	for _, resp := range responses {
+		if resp.err == nil {
+			continue
+		}
+		if agentErr, ok := resp.err.(*Error); ok && agentErr.Kind == "service-control" {
+			if errors, ok := agentErr.Value.(map[string]interface{}); ok {
+				for service, failure := range errors {
+					failures = append(failures, ServiceFailure{resp.uid, service, failure.(string)})
```

Comment:
I don't understand how we know that the failure is a string here.

Reply:
yes, that seems fragile, it will panic if it isn't

Reply:
same for the other type assertions here

Reply:
I had updated this code to check the type assertions.  I've added a test to exercise the code with various types of garbage data in the error value.